src/main/react/src/views/Feed.js

```diff
@@ -25,34 +40,68 @@ const Feed = props => {
   const [recipes, setRecipes] = useState([]);
   const [tags, setTags] = useState({});
 
+  const [page, setPage] = useState(0);
+
   const [errMsg, setErrMsg] = useState("");
 
   const [loaded, setLoaded] = useState(false);
 
-  useEffect(() => {
-    // do on feedtype change and initial render
-    let listener = app.auth().onAuthStateChanged(async user => {
-      listener();
-      let recipeData = await getRecipes(props.feedType);
-      let tagIds = {};
-      recipeData.forEach(recipe => Object.assign(tagIds, recipe.tagIds));
-      setTags(await getTags(tagIds));
+  const loadRecipes = useCallback(
+    async (user, recipePage) => {
+      let recipeData = await getRecipes(props.feedType, recipePage);
       if (user) {
-        let voteData = await getRecipesVote(recipeData);
-        recipeData.forEach((recipe, i) => (recipe.voted = voteData[i]));
-        let savedData = await getRecipesSaved(recipeData);
-        recipeData.forEach((recipe, i) => (recipe.saved = savedData[i]));
+        let [voteData, savedData] = await Promise.all([getRecipesVote(recipeData), getRecipesSaved(recipeData)]);
+        recipeData.forEach((recipe, i) => {
+          recipe.voted = voteData[i];
+          recipe.saved = savedData[i];
+        });
       }
-      setRecipes(recipeData);
+      return recipeData;
+    },
+    [props.feedType]
+  );
+
+  const [signedIn, setSignedIn] = useState(false);
+
+  let listener = app.auth().onAuthStateChanged(user => {
+    listener();
+    setSignedIn(user ? true : false);
+  });
+
+  const initRender = useCallback(async () => {
+    if (signedIn) {
+      // do on feedtype change and initial render
+      setTags(await getTags());
+      let recipeDataPage0 = await loadRecipes(app.auth().currentUser, 0);
+      setRecipes([recipeDataPage0]);
       setLoaded(true); // Forces component to re-render once recipe loading finished.
-    });
+    }
+  }, [signedIn, loadRecipes]);
+
+  useEffect(() => {
+    initRender();
     return () => {
       // return the cleanup function
       setRecipes([]);
       setTags({});
       setLoaded(false);
     };
-  }, [props.feedType]);
+  }, [loadRecipes, initRender, signedIn]);
+
+  useEffect(() => {
+    if (signedIn && loaded) {
+      if (!recipes[page + 1]) {
+        loadRecipes(app.auth().currentUser, page + 1).then(nextPageRecipes =>
+          setRecipes([...recipes, nextPageRecipes])
+        );
+      }
+    }
+    // we do not include recipes
```

Comment:
why are those lines so short? Looks like a poem!

Reply:
Fixed in eb25571